src/Avalonia.Base/Data/BindingValue.cs

```diff
@@ -0,0 +1,431 @@
+                BindingValueType.DataValidationError =>
+                    new BindingNotification(Error, BindingErrorType.DataValidationError),
+                BindingValueType.DataValidationErrorWithFallback =>
+                    new BindingNotification(Error, BindingErrorType.DataValidationError, Value),
+                _ => throw new NotSupportedException("Invalida BindingValueType."),
+            };
+        }
+
+        /// <summary>
+        /// Returns a new binding value with the specified value.
+        /// </summary>
+        /// <param name="value">The new value.</param>
+        /// <returns>The new binding value.</returns>
+        /// <exception cref="InvalidOperationException">
+        /// The binding type is <see cref="BindingValueType.UnsetValue"/> or
+        /// <see cref="BindingValueType.DoNothing"/>.
+        /// </exception>
+        public BindingValue<T> WithValue(T value)
+        {
+            if (Type == BindingValueType.DoNothing)
+            {
+                throw new InvalidOperationException("Cannot add value to DoNothing binding value.");
+            }
+
+            var type = Type == BindingValueType.UnsetValue ? BindingValueType.Value : Type;
+            return new BindingValue<T>(type | BindingValueType.HasValue, value, Error);
+        }
+
+        /// <summary>
+        /// Gets the value of the binding value if present, otherwise the default value.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public T GetValueOrDefault() => HasValue ? _value : default;
+
+        /// <summary>
+        /// Gets the value of the binding value if present, otherwise a default value.
+        /// </summary>
+        /// <param name="defaultValue">The default value.</param>
+        /// <returns>The value.</returns>
+        public T GetValueOrDefault(T defaultValue = default) => HasValue ? _value : defaultValue;
+
+        /// <summary>
+        /// Gets the value if present, otherwise the default value.
+        /// </summary>
+        /// <returns>
+        /// The value if present and of the correct type, `default(TResult)` if the value is
+        /// not present or of an incorrect type.
+        /// </returns>
+        public TResult GetValueOrDefault<TResult>()
+        {
+            return HasValue ?
+                _value is TResult result ? result : default
+                : default;
+        }
+
+        /// <summary>
+        /// Gets the value of the binding value if present, otherwise a default value.
+        /// </summary>
+        /// <param name="defaultValue">The default value.</param>
+        /// <returns>
+        /// The value if present and of the correct type, `default(TResult)` if the value is
+        /// present but not of the correct type or null, or <paramref name="defaultValue"/> if the
+        /// value is not present.
+        /// </returns>
+        public TResult GetValueOrDefault<TResult>(TResult defaultValue = default)
```

Comment:
Compiler warning for hidden overload (function above will be picked over this one when no argument has been specified so this `= default` is not needed).